### PR TITLE
Add support for compose profiles

### LINKED ISSUE #2777: Make use of profiles for compose up subset / compose tasks
https://docs.docker.com/compose/profiles/

### PR DESCRIPTION
**Pull request by bwateratmsft**
Fixes #2777.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1300,24 +1300,20 @@
...
-                                    "enum": [
-                                        "file"
-                                    ]
-                                }
+                                ]
                             }
-                        ]
+                        ],
+                        "not": {
+                            "required": [
```
Comment (bwateratmsft) [inline package.json:1312]: Cleaner way to denote one XOR the other.

File: package.json [hunk 2]
```diff
@@ -1737,6 +1742,61 @@
                     ],
                     "description": "%vscode-docker.config.template.composeUp.description%"
                 },
+                "docker.commands.composeUpSubset": {
```
Comment (bwateratmsft) [inline package.json:1745]: Only good way I could work out to deal with putting `${profileList}` toward the beginning and `${serviceList}` at the end was to make a new customizable command.

File: src/commands/compose/getComposeSubsetList.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+
+    return chosenSubsets.map(c => c.data);
+}
+
+async function getServiceSubsets(workspaceFolder: vscode.WorkspaceFolder, composeCommand: string, type: SubsetType): Promise<string[] | undefined> {
+    // TODO: if there are any profiles, then only services with no profiles show up when you query `config --services`. This makes for a lousy UX.
+
+    try {
+        // Start by getting a new command with the exact same files list (replaces the "up ..." or "down ..." with "config --services" or "config --profiles")
+        const configCommand = composeCommand.replace(composeCommandReplaceRegex, `config --${type}`);
```
Comment (bwateratmsft) [inline src/commands/compose/getComposeSubsetList.ts:113]: We should use JSON output.

File: package.json (hunk 1 above)
Comment (bwateratmsft) [reply package.json:1312]: Double check this though.

File: package.json [hunk 4]
```diff
@@ -1300,22 +1300,12 @@
                             {
                                 "required": [
```
Comment (bwateratmsft) [inline package.json:1301]: Tasks in this screenshot: incorrect (both), incorrect (neither), correct (file), correct (module)
image

File: package.json [hunk 5]
```diff
@@ -1401,22 +1404,12 @@
...
```
Comment (bwateratmsft) [inline package.json:1405]: Tasks in this screenshot: incorrect (both), incorrect (neither), correct (up), correct (down)
image

File: package.json [hunk 6]
```diff
@@ -1354,10 +1344,23 @@
...
+                                        "items": {
+                                            "type": "string"
+                                        }
+                                    },
                                     "customOptions": {
                                         "type": "string",
                                         "description": "%vscode-docker.tasks.docker-compose.dockerCompose.up.customOptions%"
                                     }
+                                },
+                                "not": {
```
Comment (bwateratmsft) [inline package.json:1359]: Tasks in this screenshot: incorrect (both), correct (services), correct (profiles), correct (neither)
image

**Review by karolz-ms — APPROVED** [review]

File: src/commands/compose/getComposeSubsetList.ts (hunk 3 above)
Comment (bwateratmsft) [reply src/commands/compose/getComposeSubsetList.ts:113]: The only way to get all the information would be to use `docker-compose config --profiles` to get a list of profiles, then use `docker-compose config --format json` once for each profile, then dedupe. IMO that's just silly, so I'll leave it as it is now.

**Pull request merged**